server/src/main/java/org/elasticsearch/reservedstate/service/ReservedStateUpdateTask.java

```diff
@@ -0,0 +1,138 @@
+    private final Collection<String> orderedHandlers;
+    private final Consumer<ErrorState> errorReporter;
+    private final ActionListener<ActionResponse.Empty> listener;
+
+    public ReservedStateUpdateTask(
+        String namespace,
+        ReservedStateChunk stateChunk,
+        Map<String, ReservedClusterStateHandler<?>> handlers,
+        Collection<String> orderedHandlers,
+        Consumer<ErrorState> errorReporter,
+        ActionListener<ActionResponse.Empty> listener
+    ) {
+        this.namespace = namespace;
+        this.stateChunk = stateChunk;
+        this.handlers = handlers;
+        this.orderedHandlers = orderedHandlers;
+        this.errorReporter = errorReporter;
+        this.listener = listener;
+    }
+
+    @Override
+    public void onFailure(Exception e) {
+        listener.onFailure(e);
+    }
+
+    ActionListener<ActionResponse.Empty> listener() {
+        return listener;
+    }
+
+    protected ClusterState execute(ClusterState state) {
+        ReservedStateMetadata existingMetadata = state.metadata().reservedStateMetadata().get(namespace);
+        Map<String, Object> reservedState = stateChunk.state();
+        ReservedStateVersion reservedStateVersion = stateChunk.metadata();
+
+        var reservedMetadataBuilder = new ReservedStateMetadata.Builder(namespace).version(reservedStateVersion.version());
+        List<String> errors = new ArrayList<>();
+
+        for (var handlerName : orderedHandlers) {
+            ReservedClusterStateHandler<?> handler = handlers.get(handlerName);
+            try {
+                Set<String> existingKeys = keysForHandler(existingMetadata, handlerName);
+                TransformState transformState = handler.transform(reservedState.get(handlerName), new TransformState(state, existingKeys));
+                state = transformState.state();
+                reservedMetadataBuilder.putHandler(new ReservedStateHandlerMetadata(handlerName, transformState.keys()));
+            } catch (Exception e) {
+                errors.add(format("Error processing %s state change: %s", handler.name(), unwrapException(e)));
+            }
+        }
+
+        if (errors.isEmpty() == false) {
+            // Check if we had previous error metadata with version information, don't spam with cluster state updates, if the
+            // version hasn't been updated.
+            if (existingMetadata != null
+                && existingMetadata.errorMetadata() != null
+                && existingMetadata.errorMetadata().version() >= reservedStateVersion.version()) {
+                logger.error("Error processing state change request for [{}] with the following errors [{}]", namespace, errors);
+
+                throw new ReservedStateVersion.IncompatibleVersionException(
+                    format(
+                        "Not updating error state because version [%s] is less or equal to the last state error version [%s]",
```

Comment:
Why is this an error? Isn't this part of the idempotent nature of the reserved state design?

Reply:
What about if the same operator settings file has an update triggered, but for the same version? Wouldn't this cause an internal error, when it should be a no-op?

Separate from this PR, I think we should consider how plugins can use this functionality without directly calling the reserved state service. As you point out, it can be error prone, so perhaps we can redesign the pluggable portion to declaratively specify their resources, and the server is actually what does the registration.

Reply:
Fair point, I throw a specific exception here that would be caught and handled differently by the caller of the service, but it's not ideal, it should be a no-op and a log message only. I'll fix this.

Reply:
> Separate from this PR, I think we should consider how plugins can use this functionality without directly calling the reserved state service. As you point out, it can be error prone, so perhaps we can redesign the pluggable portion to declaratively specify their resources, and the server is actually what does the registration.

I like this idea, perhaps though SPI we can ask all plugins if they want to declare any state chunks that need to be applied and we call the update service internally. This way no plugins will have access to the update service and we'll apply it using our rules.